File: integrations/java/oicana/src/main/java/com/oicana/CompiledDocument.java
package com.oicana;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

/**
 * A compiled document kept in memory so it can be exported on demand without re-compiling.
 *
 * Obtain one via {@link Template#compile(java.util.Map, java.util.Map, CompilationMode)}. Use
 * try-with-resources or call {@link #close()} to free the underlying document.
 *
 * <pre>{@code
 * try (var document = template.compile(inputs, blobs)) {
 *     byte[] pdf = document.exportPdf();
 *     byte[] firstPage = document.exportPng(2.0f, PageRange.single(0));
 * }
 * }</pre>
 */
public class CompiledDocument implements AutoCloseable {

    // The native document_pages output is a JSON array of {"width": <n>, "height": <n>} objects in
    // field order. The Java integration intentionally has no JSON dependency, so the fixed shape is
    // parsed directly.
    private static final Pattern PAGE_PATTERN = Pattern.compile(
            "\"width\"\\s*:\\s*([-+0-9.eE]+)\\s*,\\s*\"height\"\\s*:\\s*([-+0-9.eE]+)");

    private String documentId;
    private final List<PageSize> pages;
    private final String warnings;

    CompiledDocument(String documentId) {
        this.documentId = documentId;
        this.pages = parsePageSizes(OicanaNative.documentPages(documentId));
        this.warnings = OicanaNative.getWarnings(documentId);
    }

    /**
     * The sizes (in points) of every page, in document order.
     *
     * @return an unmodifiable list of page sizes
     */
    public List<PageSize> pages() {
        return pages;
    }

    /**
     * The number of pages in the document.
     *
     * @return the page count
     */
    public int pageCount() {
        return pages.size();
    }

    /**
     * Warnings produced by the compilation of this document.
     *
     * @return the warnings, or an empty Optional if there were none
     */
    public Optional<String> warnings() {
        return Optional.ofNullable(warnings);
    }

    /**
     * Export the whole document in the given format.
     *
     * @param exportFormat the output format
     * @return the exported document as a byte array
     */
    public byte[] export(ExportFormat exportFormat) {
        return export(exportFormat, null);
    }

    /**
     * Export the document in the given format, optionally restricted to a range of pages.
     *
     * @param exportFormat the output format
     * @param pages the 0-based, inclusive page range to export, or {@code null} for the whole
     *     document
     * @return the exported document as a byte array
     */
    public byte[] export(ExportFormat exportFormat, PageRange pages) {
        ensureOpen();
        return OicanaNative.exportDocument(
                documentId,
                exportFormat.toJsonString(),
                pages == null ? null : pages.toJsonString());
    }

    /**
     * Export the whole document to PDF.
     *
     * @return the PDF document as a byte array
     */
    public byte[] exportPdf() {
        return export(ExportFormat.pdf(), null);
    }

    /**
     * Export the document to PDF, optionally restricted to a range of pages.
     *
     * @param pages the 0-based, inclusive page range to export, or {@code null} for the whole
     *     document
     * @return the PDF document as a byte array
     */
    public byte[] exportPdf(PageRange pages) {
        return export(ExportFormat.pdf(), pages);
    }

    /**
     * Export the whole document to PNG at the default resolution of 1 pixel per point.
     *
     * @return the PNG image as a byte array
     */
    public byte[] exportPng() {
        return exportPng(1.0f, null);
    }

    /**
     * Export the whole document to PNG.
     *
     * @param pixelsPerPt resolution in pixels per point
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(float pixelsPerPt) {
        return exportPng(pixelsPerPt, null);
    }

    /**
     * Export the document to PNG at the default resolution of 1 pixel per point, optionally
     * restricted to a range of pages.
     *
     * @param pages the 0-based, inclusive page range to export, or {@code null} for the whole
     *     document
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(PageRange pages) {
        return exportPng(1.0f, pages);
    }

    /**
     * Export the document to PNG, optionally restricted to a range of pages.
     *
     * @param pixelsPerPt resolution in pixels per point
     * @param pages the 0-based, inclusive page range to export, or {@code null} for the whole
     *     document
     * @return the PNG image as a byte array
     */
    public byte[] exportPng(float pixelsPerPt, PageRange pages) {
        return export(ExportFormat.png(pixelsPerPt), pages);
    }

    /**
     * Export the whole document to SVG.
     *
     * @return the SVG document as a byte array
     */
    public byte[] exportSvg() {
        return export(ExportFormat.svg(), null);
    }

    /**
     * Export the document to SVG, optionally restricted to a range of pages.
     *
     * @param pages the 0-based, inclusive page range to export, or {@code null} for the whole
     *     document
     * @return the SVG document as a byte array
     */
    public byte[] exportSvg(PageRange pages) {
        return export(ExportFormat.svg(), pages);
    }

    /** Release the cached document. The instance must not be used after calling close(). */
    @Override
    public void close() {
        if (documentId != null) {
            OicanaNative.removeDocument(documentId);
            documentId = null;
        }
    }

    private void ensureOpen() {
        if (documentId == null) {
            throw new IllegalStateException("CompiledDocument has already been closed");
        }
    }

    private static List<PageSize> parsePageSizes(String json) {
        List<PageSize> sizes = new ArrayList<>();
        Matcher matcher = PAGE_PATTERN.matcher(json);
        while (matcher.find()) {
            double width = Double.parseDouble(matcher.group(1));
            double height = Double.parseDouble(matcher.group(2));
            sizes.add(new PageSize(width, height));
        }
        return Collections.unmodifiableList(sizes);
    }
}

File: integrations/node/oicana-node/src/CompiledDocument.ts
import {
  documentPages,
  exportDocument,
  getWarnings,
  removeDocument,
} from '@oicana/node-native';
import { type ExportFormat, Pdf, Png, Svg } from './ExportFormat.js';
import { type PageRange, serializePageRange } from './PageRange.js';

/**
 * Size of a single document page, in typographic points (pt).
 */
export interface PageSize {
  /** Page width in points. */
  width: number;
  /** Page height in points. */
  height: number;
}

/**
 * A compiled document that is kept in memory so its pages can be exported on
 * demand without re-compiling.
 *
 * Obtain one via {@link Template.compile}. Call {@link dispose} (or
 * use the `using` keyword) to release the underlying document.
 */
export class CompiledDocument implements Disposable {
  private documentId: string | undefined;

  /** Sizes (in points) of every page, in document order. */
  public readonly pages: ReadonlyArray<PageSize>;

  /**
   * Warnings produced by the compilation of this document, or `undefined` if
   * there were none.
   */
  public readonly warnings: string | undefined;

  /**
   * @internal Construct via {@link Template.compile}.
   */
  public constructor(documentId: string) {
    this.documentId = documentId;
    this.pages = JSON.parse(documentPages(documentId)) as PageSize[];
    this.warnings = getWarnings(documentId) ?? undefined;
  }

  /** Number of pages in the document. */
  public get pageCount(): number {
    return this.pages.length;
  }

  /**
   * Export the document in the given format (defaults to PDF), optionally
   * restricted to a range of pages.
   * @param format - export format specification
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public export(format: ExportFormat = Pdf, pages?: PageRange): Uint8Array {
    if (this.documentId === undefined) {
      throw new Error('CompiledDocument has already been disposed');
    }
    return exportDocument(
      this.documentId,
      JSON.stringify(format),
      serializePageRange(pages),
    );
  }

  /**
   * Export the document to PDF, optionally restricted to a range of pages.
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportPdf(pages?: PageRange): Uint8Array {
    return this.export(Pdf, pages);
  }

  /**
   * Export the document to PNG, optionally restricted to a range of pages.
   * @param pixelsPerPt - resolution in pixels per point (defaults to 1.0)
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportPng(pixelsPerPt = 1.0, pages?: PageRange): Uint8Array {
    return this.export(Png(pixelsPerPt), pages);
  }

  /**
   * Export the document to SVG, optionally restricted to a range of pages.
   * @param pages - 0-based, inclusive page range (defaults to the whole document)
   */
  public exportSvg(pages?: PageRange): Uint8Array {
    return this.export(Svg, pages);
  }

  /**
   * Release the cached document. After calling dispose() this instance must not
   * be used anymore.
   */
  public dispose(): void {
    if (this.documentId !== undefined) {
      removeDocument(this.documentId);
      this.documentId = undefined;
    }
  }

  /**
   * Enables use with the `using` keyword for automatic resource cleanup.
   */
  [Symbol.dispose](): void {
    this.dispose();
  }
}
